File: 144. Binary Tree Preorder Traversal/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        result = []
        if root != None:
            result.append(root.val)
            result.extend(self.preorderTraversal(root.left))
            result.extend(self.preorderTraversal(root.right))
        return result


class Solution2(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        stack, result = [root], []
        while len(stack) != 0 and root != None:
            n = stack.pop()
            result.append(n.val)
            if n.right != None: stack.append(n.right)
            if n.left != None: stack.append(n.left)
    
        return result

File: 144. Binary Tree Preorder Traversal/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> result;
        if ( root != NULL ) {
            result.push_back(root->val);
            vector<int> vec_left = this->preorderTraversal(root->left);
            vector<int> vec_right = this->preorderTraversal(root->right);
            result.insert(result.end(), vec_left.begin(), vec_left.end());
            result.insert(result.end(), vec_right.begin(), vec_right.end());
        }
        return result;
    }
    
};


class Solution2 {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        stack<TreeNode*> s; s.push(root);
        vector<int> result;
        
        while (!s.empty() && root) {
            TreeNode * node = s.top(); s.pop();
            result.push_back(node->val);
            if (node->right) s.push(node->right);
            if (node->left) s.push(node->left);
        }
        return result;
    }
};


